src/lib/reasoners/bitv.ml

```diff
@@ -699,28 +711,117 @@ module Shostak(X : ALIEN) = struct
 
-    let rec slice_composition eq pat (ac_eq,c_sub) = match (eq,pat) with
-      |[],[] -> (ac_eq,c_sub)
-      |st::_,n::_  when st.sz < n -> assert false
-      |st::comp,n::pt ->
-        if st.sz = n then slice_composition comp pt (st::ac_eq , c_sub)
-        else let (st_n,res,flag) = slice_var st n
-          in begin
-            match flag with
-            |Some B ->
-              let comp' = List.fold_right (fun s_t acc ->
-                  if compare_solver_simple_term s_t st <> 0
-                  then s_t::acc
-                  else st_n::res::acc
-                ) comp []
-              in slice_composition (res::comp') pt (st_n::ac_eq,c_sub)
-
-            |Some C -> let ac' = (st_n::ac_eq,(st,(st_n,res))::c_sub)
-              in slice_composition (res::comp) pt ac'
-
-            | _ -> slice_composition (res::comp) pt (st_n::ac_eq,c_sub)
-          end
-      | _ -> assert false
+    (* This is a helper function for [slice_vars].
+
+       [apply_sub_rev sub changed acc pat rcomp] applies the substitution [sub]
+       (which maps B-variable to pairs of B-variables) to the composition
+       [rcomp], which is stored in *reverse* order, and:
+
+       - Updates [changed] if the substitution effectively changes a variable
+       - Computes the resulting composition (in normal order) in [acc]
+       - Computes the resulting pattern in [pat]
+
+       If [changed', comp', pat' = apply_sub_rev sub false [] [] rcomp] then
+       the following holds:
+
+       - [pat' = List.map (fun bv -> bv.st) comp']
+       - If [changed'] is [false], [comp' = List.rev rcomp]
+
+       Note that the substitution can be recursive, so we must do a recursive
+       call on the result of applying the substitution (we can't add [v1 :: v2]
+       to [acc], we must add [v2 :: v1] to [r] instead). *)
+    let rec apply_sub_rev sub changed acc pat = function
+      | [] -> changed, acc, pat
+      | v :: r ->
+        match List.assoc v sub with
+        | v1, v2 ->
+          (* Note that this is [v2 :: v1 :: r] and not [v1 :: v2 :: r] because
+             the composition is given in reverse order. *)
+          apply_sub_rev sub true acc pat (v2 :: v1 :: r)
+        | exception Not_found ->
+          apply_sub_rev sub changed (v :: acc) (v.sz :: pat) r
+
+    (* [slice_vars eq pat (c, req, subs)] slices the variables in [eq] according
+       to the pattern [pat].  [eq] is a composition of simple terms.
+
+       If the composition [eq] does not match the pattern in [pat]:
+       - When variables in the composition are bigger than the expected pattern
+          size, the variables are split into two parts, the first of which has
+          the size expected by the pattern.
+
+          Splits involving A variables are not recorded, because A variables are
+          guaranteed to have unique occurences.
+
+          Splits involving B and C variables are accumulated in [subs].
```

Comment:
Hm, difference will probably be insignificant, but would there be a point in splitting `subs` into `b_subs` and `c_subs`? Since B variables can appear multiple times but only in a single composition, while C variables can appear multiple time but not in the same composition, I don't think that it is needed to have both substitutions in the same place.

Reply:
Unless I am mistaken/misremembering, `subs` should be a pair `(b_subs, c_subs)` for this exact reason. I could clarify the comment.

Reply:
Right, line 788